Add documentation for galley.enableAnalysis (#5216)
istio/istio.io: content/en/docs/ops/diagnostic-tools/istioctl-analyze/index.md

- **What about analysis that goes beyond configuration?**

      Today, the analysis is purely based on Kubernetes configuration, but in the future we’d like to expand beyond that. For example, we could allow analyzers to also look at logs to generate recommendations.

- **Where can I find out how to fix the errors I'm getting?**

>       Reference for configuration analysis messages can be found [here](/docs/reference/config/analysis/).

Reviewer: ```suggestion
      The set of [configuration analysis messages](/docs/reference/config/analysis/) contain descriptions of each message along with suggested fixes.
```

> ## Enabling Validation Messages for Resource Status

Reviewer: ```suggestion
## Enabling validation messages for resource status
```

*Note that this feature is experimental.*

> Starting with Istio 1.4, Galley can be set up to perform configuration analysis alongside the configuration distribution that it is primarily responsible for. Validation messages from the analysis are written to the status subresource of the affected Istio resource. This lets you find many configuration errors in your mesh without needing to run a separate command. This analysis uses the same logic and error messages as when using `istioctl analyze`.

Reviewer 2: We need to emphasize that this is completely different from 'istioctl analyze'. Specifically:

- You can run analyze against much older clusters, while this only helps with absolute newest clusters
- It only lets you get status messages one by one, so it's not at all a replacement for `analyze`, which gives you everything. In this sense, I find the statement *This lets you find many configuration errors in your mesh without needing to run a separate command* rather confusing.

Is this on by default in 1.4?

Author: It is off by default in 1.4.

Added a section with some caveats / explanations of what's different.

- **What about analysis that goes beyond configuration?**

      Today, the analysis is purely based on Kubernetes configuration, but in the future we’d like to expand beyond that. For example, we could allow analyzers to also look at logs to generate recommendations.

- **Where can I find out how to fix the errors I'm getting?**

>       The set of [configuration analysis messages](/docs/reference/config/analysis/) contain descriptions of each message along with suggested fixes.

Reviewer: ```suggestion
      The set of [configuration analysis messages](/docs/reference/config/analysis/) contains descriptions of each message along with suggested fixes.
```

## Enabling validation messages for resource status

> *Note that this feature is experimental.*

Reviewer: ```suggestion
{{< boilerplate experimental-feature-warning >}}
```

Starting with Istio 1.4, Galley can be set up to perform configuration analysis alongside the configuration distribution that it is primarily responsible for, via the `galley.enableAnalysis` flag.
This analysis uses the same logic and error messages as when using `istioctl analyze`. Validation messages from the analysis are written to the status subresource of the affected Istio resource.

> e.g. if you have a misconfigured gateway on your "ratings" virtual service, running `kubectl get virtualservice ratings` would give you something like:

Reviewer: ```suggestion
For example. if you have a misconfigured gateway on your "ratings" virtual service, running `kubectl get virtualservice ratings` would give you something like:
```

> {{< text plain >}}

Reviewer: ```suggestion
{{< text yaml >}}
```

apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"networking.istio.io/v1alpha3","kind":"VirtualService","metadata":{"annotations":{},"name":"ratings","namespace":"default"},"spec":{"hosts":["ratings"],"http":[{"route":[{"destination":{"host":"ratings","subset":"v1"}}]}]}}
  creationTimestamp: "2019-09-04T17:31:46Z"
  generation: 11
  name: ratings
  namespace: default
  resourceVersion: "12760039"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/default/virtualservices/ratings
  uid: dec86702-cf39-11e9-b803-42010a8a014a
spec:
  gateways:
  - bogus-gateway
  hosts:
  - ratings
  http:
  - route:
    - destination:
        host: ratings
        subset: v1
status:
  validationMessages:
  - code: IST0101
    level: Error
    message: 'Referenced gateway not found: "bogus-gateway"'
{{< /text >}}

`enableAnalysis` runs in the background, and will keep the status field of a resource up to date with its current validation status. Note that this isn't a replacement for `istioctl analyze`:

> - Not all resources have a custom status field (e.g. Kubernetes *namespace* resources), so messages attached to those resources won't show validation messages.

Reviewer: ```suggestion
- Not all resources have a custom status field (e.g. Kubernetes `namespace` resources), so messages attached to those resources won't show validation messages.
```

> - `enableAnalysis` only works on Istio versions starting with 1.4, while `istioctl analyze` can be used on older cluster versions.

Reviewer: ```suggestion
- `enableAnalysis` only works on Istio versions starting with 1.4, while `istioctl analyze` can be used with older versions.
```

- While it makes it easy to see what's wrong with a particular resource, it's harder to get a holistic view of validation status in the mesh.

> To enable this feature with `helm template`:

Reviewer: ```suggestion
You can enable this feature with:
```

{{< text bash >}}
$ helm template install/kubernetes/helm/istio --name istio --namespace istio-system \
    --set galley.enableAnalysis=true | kubectl apply -f -
{{< /text >}}

> Or with `helm install`:

Reviewer: ```suggestion
Or with:
```
